### PR TITLE
 plex-desktop: init at 1.96.0

### LINKED ISSUE #252131: Package request: plex-desktop
**Project description**

Streaming media player for Plex

**Metadata**

* homepage URL: https://www.plex.tv/fr/media-server-downloads/#plex-app
* source URL: https://snapcraft.io/plex-desktop
* license: proprietary
* platforms: linux

---

Unfortunately, this app is only available as a Snap or a Flatpak. It looks like Snap is easier to unpack, as we have [one example with Spotify](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/audio/spotify/linux.nix).

There is also [an AUR package](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=plex-desktop), which appears to be unmaintened.

### PR DESCRIPTION
**Pull request by detroyejr**
## Description of changes

plex-desktop: init at 1.95.3

Based on the [spotify](https://github.com/NixOS/nixpkgs/blob/d0f207da0e472ca4bd1de8e14f827ab92b1fdcd0/pkgs/applications/audio/spotify/linux.nix) derivation which pulls from the snap store. Also borrowed the excellent update script from there as well.

Fixes https://github.com/NixOS/nixpkgs/issues/252131

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Updated TODO for 2024-07-11:

- [ ] Downloads
- [ ] Investigate Plex & Plex,sh processes hang around after closing the app. Other apps do this so this though, so maybe this isn't specific to Plex.

---

Original Post

Tested this on 3 different NixOS systems and everything generally works.

* Gnome (w/ integrated Intel GPU) - everything works & always launches successfully.
* Hyprland (w/ integrated intel GPU) - sometimes fails to launch for an unknown reason. Messing with `QT_QPA_PLATFORM=xcb` generally works.
* Hyprland (w/ Nvidia) - needs [`nvidia-offload`](https://nixos.wiki/wiki/Nvidia) environment variables set (see below). Works flawlessly.

```
export __NV_PRIME_RENDER_OFFLOAD=1
export __NV_PRIME_RENDER_OFFLOAD_PROVIDER=NVIDIA-G0
export __GLX_VENDOR_LIBRARY_NAME=nvidia
export __VK_LAYER_NV_optimus=NVIDIA_only
```

The steam derivation provide an `extraEnv` option which seems like a reasonable solution. In my own configuration, I just override those as needed for this one particular machine. This is probably what I'll do unless someone someone has a better suggestion.

```
# Override the package passed to programs.steam.package based on the isNvidia flag in my config.
pkgs.steam-small.override
{
  extraEnv =
    if isNvidia
    then {
       __NV_PRIME_RENDER_OFFLOAD = 1;
       __NV_PRIME_RENDER_OFFLOAD_PROVIDER = "NVIDIA-G0";
       __GLX_VENDOR_LIBRARY_NAME = "nvidia";
       __VK_LAYER_NV_optimus = "NVIDIA_only";
    }
    else {};
};
```

I'd appreciate any feedback on this that would make this launch more reliably.

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by detroyejr** [comment]
Environment variable overrides now work as follows:
```
plex-desktop.override{ 
  extraEnv = {
    __NV_PRIME_RENDER_OFFLOAD = 1;
    __NV_PRIME_RENDER_OFFLOAD_PROVIDER = "NVIDIA-G0";
    __GLX_VENDOR_LIBRARY_NAME = "nvidia"
    __VK_LAYER_NV_optimus = "NVIDIA_only"; 
  };
 }
```

**Comment by Redhawk18** [comment]
I'm a maintainer of plexamp let me know when you need approval

**Comment by detroyejr** [comment]
@Redhawk18 Thanks! A couple parts that seem to be broken:

- Downloads
- Rentals
- DRM content

There's a sqlite error on launch which could be related. It's correctly writing to `~/.cache` and `~/.local/share/plex` and the only database I can see (com.plexapp.plugins.library.db) is writable so I haven't figured out what exactly it needs.

Login, personal server content, live tv, and some of their ad-supported content all works. It's good enough that I'd be fine with merging this as-is and working out the other issues down the road. Let me know what you think.

**Comment by Redhawk18** [comment]
I'm unsure how to fix these errors, there's not much I can personally do.

**Comment by detroyejr** [comment]
No worries. Not expecting you to put any time in just noting what doesn't work. If this is good enough for whoever has merge rights, I'd say let's get it in and I'll keep trying to figure out where the problem lies and issue another PR then.

**Comment by Redhawk18** [comment]
> No worries. Not expecting you to put any time in just noting what doesn't work. If this is good enough for whoever has merge rights, I'd say let's get it in and I'll keep trying to figure out where the problem lies and issue another PR then.

I believe you'll have to post in discord because only nixos team members have perms to merge new packages. Also once it's merge we should remove the much older and outdated version of plex desktop.

**Comment by detroyejr** [comment]
While testing 4k/HD content, I found Plex wasn't using my integrated intel GPU. Thanks to NixGL, I was able to narrow down the problem to the version of `libglvdn`. It needs this library to work properly, but the newest version in unstable isn't compatible with the libraries that ship in the snap distribution. Pinning this to a previous version seems to have done the trick. This also solved the stability issues I was experiencing on system no. 3 so hopefully everything is a little more stable.

File: pkgs/by-name/pl/plex-desktop/package.nix [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        'Icon=''${SNAP}/meta/gui/icon.png' \
+        'Icon=${plex-desktop}/meta/gui/icon.png'
+    '';
+
+    runScript = writeShellScript "plex-desktop.sh" ''
+      # Widevine won't download unless this directory exists.
+      mkdir -p $HOME/.cache/plex/
+
+      set -o allexport
+      LD_LIBRARY_PATH=:${plex-desktop}:${plex-desktop}/lib:${plex-desktop}/usr/lib/x86_64-linux-gnu:${lib.makeLibraryPath [ libglvnd-1_4_0 ]}"''${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}"
```
[the remark this answers is not in the document]
Comment (detroyejr) [reply pkgs/by-name/pl/plex-desktop/package.nix]: Unsure exactly when using ${plex-desktop} for library paths is acceptable, so I've replaced both ${plex-desktop}/lib and ${plex-desktop}/usr/lib with:

```
PLEX_USR_PATH=${lib.makeSearchPath "usr/lib/x86_64-linux-gnu"  [ plex-desktop ]}
...
LD_LIBRARY_PATH=${lib.makeLibraryPath [ plex-desktop libglvnd-1_4_0 ]}:$PLEX_USR_PATH
LIBGL_DRIVERS_PATH=$PLEX_USR_PATH/dri
```

**Comment by detroyejr** [comment]
@SuperSandro2000 Thank you for taking the time to review this! I appreciate the work you and everyone on the team does to keep nix going.

I've gone through and accepted most of your suggested changes and made some additions based on your comments. Let me know if I can do anything else.

File: pkgs/by-name/pl/plex-desktop/package.nix [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+in
+buildFHSEnv {
+    name = "${pname}-${version}";
+    targetPkgs = pkgs: [ xkeyboard_config ];
+
+    extraInstallCommands = ''
+      mkdir -p $out/share/applications $out/share/icons/hicolor/scalable/apps
+      install -m 444 -D ${plex-desktop}/meta/gui/plex-desktop.desktop $out/share/applications/plex-desktop.desktop
+      substituteInPlace $out/share/applications/plex-desktop.desktop \
+        --replace-warn \
```
Comment (SuperSandro2000) [inline pkgs/by-name/pl/plex-desktop/package.nix:131]:
```suggestion
        --replace-fail \
```

**Comment by Supergamer1337** [comment]
I don't know if this is the correct place to ask this (and please show me where I should go if it is not), but it does not appear that the sign in works? Is this a known issue, or just something on my machine? Is there a configuration option that needs to be added for sign in button to correctly open my web browser perhaps?

**Comment by detroyejr** [comment]
I meant to make a note of this in the PR. I think this is an issue for other applications as well and depends on the window manager you use. 

I needed this in my configuration.nix for Hyprland at least, not sure if my machine w/ Gnome needed any tweaking at all.
```
# Plex needs this to login/click on links.
xdg.portal = {
   enable = true;
   extraPortals = [ pkgs.xdg-desktop-portal-gtk ];
   xdgOpenUsePortal = true;
 };
```

**Comment by jervw** [comment]
I've been testing the package for nearly a week and have encountered two minor issues. First, CJK characters are not rendering at all. Second, I'm wondering if native Wayland support is possible. I haven't tested the upstream Snap or Flatpak packages to see if they currently support Wayland at all.

**Comment by detroyejr** [comment]
I'm unsure about solving the 1st issue. I can take a look, but if you have any suggestions please let me know.

For issue 2, there's a line in the Plex.sh startup script that comes with the snap:
```
# Disable wayland unless the user has set the value before
export DISABLE_WAYLAND=${DISABLE_WAYLAND:-1}
```
So, they disable by default which suggests to me that they don't support Wayland natively? I'm unsure how to confirm this definitively though.

**Pull request merged**